pytorch_lightning/loops/training_epoch_loop.py

```diff
@@ -0,0 +1,403 @@
+        self.is_last_batch = is_last
+
+        # ------------------------------------
+        # TRAINING_STEP + TRAINING_STEP_END
+        # ------------------------------------
+        with self.trainer.profiler.profile("run_training_batch"):
+            batch_output = self.batch_loop.run(batch, self.iteration_count, self._dataloader_idx)
+            self.batches_seen += 1
+
+        # when returning -1 from train_step, we end epoch early
+        if batch_output.signal == -1:
+            raise StopIteration
+
+        batch_end_outputs = [opt_idx_out for opt_idx_out in batch_output.training_step_output if len(opt_idx_out)]
+        processed_batch_end_outputs = self._prepare_outputs(batch_end_outputs, batch_mode=True)
+
+        # hook
+        self.trainer.call_hook(
+            'on_train_batch_end', processed_batch_end_outputs, batch, self.iteration_count, self._dataloader_idx
+        )
+        self.trainer.call_hook('on_batch_end')
+        self.trainer.logger_connector.on_batch_end()
+
+        # figure out what to track for epoch end
+        self.track_epoch_end_reduce_metrics(self.epoch_output, batch_end_outputs)
+
+        # -----------------------------------------
+        # SAVE METRICS TO LOGGERS AND PROGRESS_BAR
+        # -----------------------------------------
+        self.trainer.logger_connector.update_train_step_metrics()
+
+    def on_advance_end(self):
+        """Runs validation and Checkpointing if necessary.
+
+        Raises:
+            StopIteration: if :attr:`done` evaluates to ``True`` to finish this epoch
+        """
+        # -----------------------------------------
+        # VALIDATE IF NEEDED + CHECKPOINT CALLBACK
+        # -----------------------------------------
+        should_check_val = self.should_check_val_fx(self.iteration_count, self.is_last_batch)
+        if should_check_val:
+            self.trainer.validating = True
+            self.trainer._run_evaluation()
+            self.trainer.training = True
+
+        # -----------------------------------------
+        # SAVE LOGGERS (ie: Tensorboard, etc...)
+        # -----------------------------------------
+        self.save_loggers_on_train_batch_end()
+
+        # update LR schedulers
+        self.update_lr_schedulers('step')
+        self.trainer.checkpoint_connector.has_trained = True
+
+        self.total_batch_idx += 1
+
+        # progress global step according to grads progress
+        self.increment_accumulated_grad_global_step()
+
+        if self.done:
+            raise StopIteration
+
+    def on_run_end(self) -> List[List[STEP_OUTPUT]]:
+        """Calls the on_epoch_end hook.
+
+        Returns:
+            The output of each training step for each optimizer
+
+        Raises:
+            MisconfigurationException: ``train_epoch_end`` does not return ``None``
+        """
+        if self.batches_seen == 0:
+            # dataloader/iterator did not produce a batch
+            return
```

Comment:
@awaelchli do you remember why this was necessary?

Reply:
yes, basically if the epoch is empty we don't want to run epoch_end hooks or invoke any logging. when does this happen? a user can stop training or skip an epoch by returning a signal from one of the hooks. I don't remember the exact use case. Pretty sure if you remove this line a test will fail and inform you exactly the reason.